### DISCUSSION TITLE
How Sampler creates one single Sample from multiple collected stat?

### DISCUSSION THREAD

**Discussion by rostamiani** [ObserveRTC/observer#97] [Q&A] [answered]
As I understood, there is a collector that collects stats. And a sampler that makes samples from multiple collected stats. Is it correct?

My question is: How it's making a single sample? Does it average some data or using max/min? or just the last stat is used?
For example with this configs:
```js
  monitor = createClientMonitor({
    collectingPeriodInMs: 3000,
    samplingPeriodInMs: 10000,
    sendingPeriodInMs: 20000,
```
After 10s the first Sample is created but there are 3 different collected stats! Does Sampler combine these three collections?

Thanks

**Comment by balazskreith** [comment, marked as answer]
**Short answer:**
no aggregation, the collecting period and sampling period are separated for different reasons. 
A Sample is a snapshot of the stats at the moment it is invoked. 

In your example what happens is that the sampler creates a Sample after 10s, 20s, which means it makes a snapshot after the 3rd, 6th, and 9th collected stats. After 20s it will send 2 Samples, and after 40s it sends 2 Samples again.

**Longer reply:**
The reason for the separation is to decide what the granuality you want to work on the client-side and what is the granuality you want to save at the server side. 
Most or almost every measurement is accumulated in the stats (bytesReceived, packetsReceived, etc), so for those calculating average is pointless. 
If you want to detect let's say congestion at the client-side you can make the collecting period set to between 500ms - 2000ms, and put some detectors on the monitor like:
```javascript
monitor.events.onStatsCollected(() => {
  const storage = monitor.storage;
  let congested = false;
  for (const outboundRtp of storage.outboundRtps()) {
    if (myAwesomeDetector(outboundRtp)) congested = true
  }
  if (congested) {
    monitor.sample();
  }
});
```
In this case you detect congestion with your `myAwesomeDetector` method and if a congestion is detected you order the monitor to make a sample at that moment, which is out of the regular sampling period, but you can do that, so you will have the snapshot of the stats at the server when congestion is detected.
